feat: add new invokeSolc hook

kanej opened the pull request:
Add a `invokeSolc` hook that is called to run the `solc` compiler that Hardhat has selected for a given compilation job. The hook passes in:

* the Compiler instance that Hardhat has selected for the compilation job
* the Solc input generated from the compilation job - presenting an extension point for transforming the input
* the Solc config that was used by Hardhat to load the particular Solc compiler

It returns the Solc output from running `solc`.

The `invokeSolc` provides an extension point that can be used to transform or manipulate the Solc input before the running of `solc`, or similarly to manipulate the output from the compiler.

## invokeSolc Hook

```ts
/**
 * Hook triggered to invoke a passed in Solc compiler on the
 * Solc input generated for a given compilation job.
 * This hook allows for manipulating the Solc input passed into the Solc
 * compiler Hardhat has selected for the compilation job, and similarly to
 * manipulate the Solc output.
 *
 * @param context The hook context.
 * @param compile The Solc compiler selected by Hardhat for this compilation
 * job.
 * @param solcInput The solc input json constructed from the compilation
 * job.
 * @param solcConfig The configuration used to setup solc e.g. version.
 * @param next A function to call the next handler for this hook, or the
 * default implementation if no more handlers exist.
 */
invokeSolc(
  context: HookContext,
  compiler: Compiler,
  solcInput: CompilerInput,
  solcConfig: SolcConfig,
  next: (
    nextContext: HookContext,
    nextCompiler: Compiler,
    nextSolcInput: CompilerInput,
    nextSolcConfig: SolcConfig,
  ) => Promise<CompilerOutput>,
): Promise<CompilerOutput>;
```

Resolve #7646

kanej committed:
feat: add new onSolcCompileComplete hook

Add a hook that is called immediately after compilation passing the solc
input/output plus the solc config.

This enables generation of additional files and artifacts based on the
compilation. Note solc may have raised compilation errors/warnings -
that will still constitute a build that will trigger.

Resolves #7646.

afa7789 commented:
up

kanej committed:
feat: rework to `invokeSolc` hook

Switch from compile complete event hook to an action based hook.

kanej commented on v-next/hardhat/test/internal/builtin-plugins/solidity/hooks.ts:
```diff
@@ -0,0 +1,110 @@
+  describe("invokeSolc", () => {
+    useFixtureProject("solidity/simple-project");
+
+    const expectedSolidityVersion = "0.8.23";
```
I am controlling the Solidity version here to let me check for it against the passed in Solc compiler instance.

alcuadrado reviewed (approved):
pre-approving, but take a look at the comment i left

kanej committed:
fix: use the hooks provided versions of compiler/input

Overriding in a hook won't work unless the versions used are those
passed from the previous hook in the chain.
